lib/Dialect/Torch/Transforms/RefineTypes.cpp

```diff
@@ -923,6 +930,38 @@ ChangeResult TypeAnalyzer::visitAtenSqueezeOp(
   return getLatticeElement(op.getResult()).join(knowledge);
 }
 
+ChangeResult TypeAnalyzer::visitAtenNllLossForwardOp(
+    AtenNllLossForwardOp op,
+    ArrayRef<LatticeElement<ValueKnowledge> *> operands) {
+  auto self = operands[0]->getValue();
+  auto target = operands[1]->getValue();
+  auto weight = operands[2]->getValue();
+  auto knowledge =
+      ValueKnowledge::getNotNonePessimisticValueState(op.getContext());
+
+  // Contains Knowledge of shape and dtype for the 1st result.
+  knowledge.dtype = self.dtype;
+  int64_t reduction;
+  unsigned resultRank = self.sizes.size();
+
+  // Contains Knowledge of shape and dtype for the 2st result.
+  auto knowledge2 =
+      ValueKnowledge::getNotNonePessimisticValueState(op.getContext());
+  knowledge2.dtype = self.dtype;
+  knowledge2.sizes.resize(0, kUnknownSize);
+  knowledge2.hasSizes = true;
+
+  if (self.hasSizes &&
+      matchPattern(op.reduction(), m_TorchConstantInt(&reduction))) {
+    resultRank = reduction == 1 ? resultRank - 1 : resultRank;
+  }
+  knowledge.sizes.resize(resultRank-1, kUnknownSize);
```

Comment:
I think the `outputKnowledge` can be more accurate. Use `N*C*d0*d1` as an example. For none reduction case, it's the C dimension reduced. For reduction case, it's the MiniBatch(N) dimension and C dimension both got reduced.

Reply:
https://github.com/pytorch/pytorch/blob/e0fb228e0346a01f16592a1ced106e0f7d5efafd/aten/src/ATen/native/LossNLL.cpp#L24  The self dim is either 1 or 2.